lsp/source/server.mts

```diff
@@ -370,28 +370,23 @@ documents.onDidClose(({ document }) => {
 
 documents.onDidOpen(async ({ document }) => {
   console.log("open", document.uri)
-  const sourcePath = documentToSourcePath(document)
-  assert(sourcePath)
-
-  const service = await ensureServiceForSourcePath(sourcePath)
-  if (!service) return
-
-  service.host.addOrUpdateDocument(document)
 })
 
+// Buffer up changes to documents so we don't stack transpilations and become unresponsive
+const changeQueue = new Set<TextDocument>()
+setInterval(() => {
+  if (changeQueue.size === 0) return
+  Array.from(changeQueue).forEach((document) => {
+    updateDiagnostics(document)
+  })
+  changeQueue.clear()
+}, 50)
```

Comment:
That's essentially what I tried to start with but then found out that when the compile was running and taking ~400ms all change events were halted (since it's single threaded). Then the next change event would process, take ~400ms, halting the next change event, etc.

So it looks like we'll need some kind of event buffering, offloading the work to a background thread or both.

Thanks for taking a look at this! I'll see if I can better document the LSP building/testing process so hopefully it will be easier for people to contribute.

Reply:
You should be able to use `setTimeout(processQueue, 0)` or `queueMicrotask(processQueue)` to schedule processing the queue after all accumulated events get added to the queue.

Unless there are actual delays from getting the events over the network, in which case we could use a debounce approach (only do work when no new events for 50ms, or we've been receiving events for more than 1sec or something).

Or we could use another thread like you say, but workers are a bit limited in what they can do...